### ISSUE TITLE
High WebSocket memory usage in Jetty 10

### ISSUE THREAD

**Issue by SerCeMan** [eclipse/jetty.project#6696]
**Jetty version(s)**

`10.0.6`

**Java version/vendor** `(use: java -version)`

```
$ java -version
openjdk version "11.0.12" 2021-07-20 LTS
OpenJDK Runtime Environment Zulu11.50+19-CA (build 11.0.12+7-LTS)
OpenJDK 64-Bit Server VM Zulu11.50+19-CA (build 11.0.12+7-LTS, mixed mode)
```

**OS type/version**

Mac/Ubuntu

**Description**

Hi, Jetty maintainers!

I got back to trying to reproduce https://github.com/eclipse/jetty.project/issues/6328. While I wasn't able to reproduce the referenced issue yet, I seem to have found the main issue that we experienced with the upgrade to Jetty 10, which was related to memory usage. 

Compared to Jetty 9, the memory usage per open WebSocket connection has increased significantly, making it problematic for using when servers maintain a large number of connections. I've put together a minimal reproducer project here: [jetty10-memory-test](https://github.com/SerCeMan/jetty10-memory-test), which implements two minimal WebSocket echo servers on Jetty 9, and Jetty 10, both run with `Xmx1G` and a test that opens 10k connections. 

When running the tests against a bare-bones jetty 9 server, the memory usage does not increase much. 
<img width="1208" alt="Screen Shot 2021-09-05 at 12 13 56 pm" src="https://user-images.githubusercontent.com/1780970/132112642-a396582f-3281-4260-b016-6504d2f42f55.png">

However, when running the tests against Jetty 10, the memory usage grows significantly, and the server dies with an out of memory error after reaching ~9k connections. 
<img width="1202" alt="Screen Shot 2021-09-05 at 12 15 31 pm" src="https://user-images.githubusercontent.com/1780970/132112674-520f920c-a243-4d6f-8c54-a74b02503718.png">

The OOM heapdump seems to match the heapdump from the load tests which discovered the issue, and it hints at `org.eclipse.jetty.websocket.core.server.internal.RFC6455Negotiation`.
<img width="1062" alt="Screen Shot 2021-09-05 at 12 19 25 pm" src="https://user-images.githubusercontent.com/1780970/132112733-3ea2d623-a8d1-4248-a89d-6992308111bc.png">

A heap dump can also be taken by uncommenting a line in [`start.sh`](https://github.com/SerCeMan/jetty10-memory-test/blob/master/start.sh#L13). The heapdump points at the fact that `RFC6455Negotiation` is referenced from a number of places through `upgradeRequest` and `upgradeResponse`. 

<img width="1888" alt="Screen Shot 2021-09-05 at 12 33 01 pm" src="https://user-images.githubusercontent.com/1780970/132112976-18aa028f-7332-4b41-bf0a-d72596423523.png">

When looking at a single instance of `RFC6455Negotiation`, one can notice that the biggest memory consumer is the `baseRequest`. 
<img width="1023" alt="Screen Shot 2021-09-05 at 12 32 31 pm" src="https://user-images.githubusercontent.com/1780970/132113004-04c143c5-8be5-4321-9822-175193e1c6a9.png">

One potential fix would be to set `baseRequest` to `null` after the request upgrade, for instance at the end of `org.eclipse.jetty.websocket.core.server.internal.AbstractHandshaker#upgradeRequest`. As far as I can tell, it's the only reference. However, I haven't tested the hypothesis. I'm not sure if that's the best solution, but happy to raise a PR if it sounds good to you. Thanks!

**How to reproduce?**

I've put together a README that describes running the tests here: https://github.com/SerCeMan/jetty10-memory-test.

**Comment by joakime** [comment]
I'm a bit confused.
You state the issue is in version `10.0.6`, but then reference the work in issue #6328.

Are you using 10.0.6 (public release)? or 10.0.7-SNAPSHOT (from snapshots repo)? or are you building and testing the `jetty-10.0.x` branch HEAD?

**Comment by SerCeMan** [comment]
Hi, @joakime! As far as I can tell, the issue affects all versions on the `10.x` branch, the reproducer is built on the latest stable `10.x` release, `10.0.6`, [build.gradle#L17-L22](https://github.com/SerCeMan/jetty10-memory-test/blob/670730027d2c913f61705994810cc44d713d13d1/jetty10/build.gradle#L17-L22).

**Comment by lachlan-roberts** [comment]
Interestingly enough the test at https://github.com/SerCeMan/jetty10-memory-test fails on `10.0.6` but not on the latest `10.0.x` branch. It now appears to perform the same as 9.4 in this test. This may be due to the improvements from PR #6635.

Either way we can cleanup the `WebSocketNegotiation` to not store the original `Request` object after the upgrade.

**Comment by lachlan-roberts** [comment]
Closing as this seems to be fixed in `10.0.x` and PR #6698 has been merged to remove the reference to the original `Request` object.
